### PR TITLE
[ENH] speed up signed_weights and gamma calculations in expgrad

### PR DESCRIPTION
**Pull request by MiroDudik**
Signed-off-by: Miro Dudik <mdudik@gmail.com>

## Description
This replaces calculations of `signed_weights` and `gamma` in `UtilityParity` by matrix multiplication. For simple estimators (like logistic regression) this can lead to substantial speed-ups.

## Tests

- [x] no new tests required
- [ ] new tests added
- [ ] existing tests adjusted

## Documentation

- [x] no documentation changes needed
- [ ] user guide added or updated
- [ ] API docs added or updated
- [ ] example notebook added or updated

### CODE REVIEW

**Comment by MiroDudik** [comment]
@romanlutz -- i wonder whether this will help with the time-out in #1008.

File: fairlearn/reductions/_moments/utility_parity.py [hunk 1]
```diff
@@ -165,6 +165,34 @@ def load_data(
         self.index = signed.index
         self.default_objective_lambda_vec = None
 
+        # Fill information about the matrix U, which is used to calculate signed weights
+        # and gamme vector. The matrix has a row corresponding to each example and column
```
Comment (romanlutz) [inline fairlearn/reductions/_moments/utility_parity.py:169]:
```suggestion
        # and gamma vector. The matrix has a row corresponding to each example and column
```

**Review by romanlutz — APPROVED** [review]
Thanks for providing extensive comments.

File: fairlearn/reductions/_moments/utility_parity.py [hunk 2]
```diff
@@ -165,6 +165,34 @@ def load_data(
...
+        # gamma = -U.T.dot(utilities[y_pred]) / total_samples
+        #
+        # If the i-th example's event and group are "e" and "g" then the entries in the
+        # i-th row of the matrix, indexed by tuples (sign, event, group), are defined as:
+        #
+        # U_i,(+,e',g') =      1[e=e']/P(e) - rho*1[e=e',g=g']/P(g,e)
+        # U_i,(-,e',g') = -rho*1[e=e']/P(e) +     1[e=e',g=g']/P(g,e)
+        #
+        # This corresponds to the expression given right above Theorem 1 of the paper
+        # "A Reductions Approach to Fair Classification".
```
[the remark this answers is not in the document]
Comment (MiroDudik) [reply fairlearn/reductions/_moments/utility_parity.py]: sounds good. done.

**Comment by iofall** [comment]
> @romanlutz -- i wonder whether this will help with the time-out in #1008.

Removing the `apply` call should help speed things up. But I am not sure if it will help with the PR 1008. From the runs I have performed - [Link](https://app.circleci.com/pipelines/github/Temp-FL/fairlearn/16/workflows/c6ce1b5a-59a0-48eb-8ad5-e82d10a142a1/jobs/34), the slowdown occurs in the `fit` call. The estimator takes a long time to fit on `circleci`, but does so smoothly on local.

Maybe we can try pickling / use skops on the local fit and use this saved model while building docs on `circleci`.

**Comment by MiroDudik** [comment]
> > @romanlutz -- i wonder whether this will help with the time-out in #1008.
> 
> Removing the `apply` call should help speed things up. But I am not sure if it will help with the PR 1008. From the runs I have performed - [Link](https://app.circleci.com/pipelines/github/Temp-FL/fairlearn/16/workflows/c6ce1b5a-59a0-48eb-8ad5-e82d10a142a1/jobs/34), the slowdown occurs in the `fit` call. The estimator takes a long time to fit on `circleci`, but does so smoothly on local.
> 
> Maybe we can try pickling / use skops on the local fit and use this saved model while building docs on `circleci`.

Interesting! From the print-out I cannot tell whether the time is spent inside clone() or inside estimator.fit(). Can you separate cloning from fitting? I am still puzzled by the discrepancy between local and circleci.

**Review by iofall — COMMENTED** [review]
https://app.circleci.com/pipelines/github/Temp-FL/fairlearn/17/workflows/7728b8cb-95d8-4926-a1c4-c0a1b111f10f/jobs/35

This run should be clearer. I have put logs around - `signed_weights`, `cloning` and `fit`.

**Comment by MiroDudik** [comment]
> https://app.circleci.com/pipelines/github/Temp-FL/fairlearn/17/workflows/73731d85-5a59-432f-afc8-c5ef7f77b37a/jobs/37
> 
> This run should be clearer. I have put logs around - `signed_weights`, `cloning` and `fit`.

Thanks for narrowing it down!

I did a bit poking around and it seems that lightgbm hangs in some parallel execution environments: https://lightgbm.readthedocs.io/en/latest/FAQ.html#lightgbm-hangs-when-multithreading-openmp-and-using-forking-in-linux-at-the-same-time

So restricting lightgbm to one thread fixes the issue. The documentation now builds in 4 minutes:
https://app.circleci.com/pipelines/github/fairlearn/fairlearn/3398/workflows/fc3b6ff5-9c0f-4e6d-bc6e-1ef0fbdfcd8b/jobs/6664

**Comment by iofall** [comment]
> Thanks for narrowing it down!

Glad that helped! 
 
> So restricting lightgbm to one thread fixes the issue. The documentation now builds in 4 minutes: https://app.circleci.com/pipelines/github/fairlearn/fairlearn/3398/workflows/fc3b6ff5-9c0f-4e6d-bc6e-1ef0fbdfcd8b/jobs/6664

Awesome, that's really nice 🔥

**Review by riedgar-ms — APPROVED** [review]

**Pull request merged**